service/routes.py

```diff
@@ -287,6 +287,69 @@ def create_items(shopcart_id):
         jsonify(new_item.serialize()), status.HTTP_201_CREATED
         )
 
+######################################################################
+# UPDATE AN ITEM
+######################################################################
+@app.route("/shopcarts/<int:shopcart_id>/items/<int:item_id>", methods = ["PUT"])
+def update_items(shopcart_id, item_id):
+    """
+    Update an existing item in shopcart {shopcart_id}.
+    Update quantity and/or price
+
+    Args:
+        shopcart_id (int): The shopcart to be updated
+        body of API call (JSON): 
+            quantity (int) 
+            price (float/int)
+
+            have either quantity or price, or both
+
+    Returns:
+        status code: 201 if successful, 
+        404 if the requested shopcart_id or item_id does not exist,
+        400 if data type errors.
+        message (JSON): new item if successful, otherwise error messages
+    """
+    app.logger.info("Request to update an item")
+    check_content_type("application/json")
+    req = request.get_json()
+    if not "quantity" in req.keys() and not "price" in req.keys():
+        abort(status.HTTP_400_BAD_REQUEST, "Must have either quantity or price.")
+    if "quantity" in req.keys():
+        if not isinstance(req["quantity"], int) or req["quantity"] < 0:
+            abort(status.HTTP_400_BAD_REQUEST, "Invalid quantity.")
+        else:
+            quantity = req["quantity"]
+    if "price" in req.keys():
+        if (not isinstance(req["price"], int) and not isinstance(req["price"], float)) or req["price"] < 0:
+            abort(status.HTTP_400_BAD_REQUEST, "Invalid price.")
+        else:
+            price = req["price"]
+    
+    # Make sure the shopcart exists
+    shopcart = Shopcart.find_shopcart(shopcart_id)
+    if not shopcart:
+        abort(status.HTTP_404_NOT_FOUND, "Shopcart with id {shopcart_id} was not found.")
+    # Make sure the item exists
+    item = Shopcart.find_item(shopcart_id, item_id)
+    if not item:
+        abort(status.HTTP_404_NOT_FOUND, "item with id {item_id} was not found.")
+    
+    # Now proceed to update
+    if quantity:
```

Comment:
I don't understand this line and the if price below. Why do we need these 2 extra ifs to check? In particular, I think this implementation will fail if the quantity or the price is 0 - in my implementation for create_items, I assumed quantity cannot be 0 for creation but price can be 0; for updating items I think we should allow the same. (Probably shouldn't allow quantity to be 0 as that would be a deletion and PUT should not delete resources). But those checks should already be carried out during data type checking above, and these 2 ifs can be deleted right?

Reply:
Oh wait I see maybe only one of them is set

Reply:
I understand the test passes but I'm confused about this syntax. Is this testing whether quantity has been set as a local variable? Because when I try to do in a test scripy ```if foo: pass ``` without foo being set, I get a NameError. So if there is only quantity in the request, surely ```if price``` would throw an error since we haven't set a variable named price. Sorry I may be missing something very obvious due to my inexperience.

Reply:
Thanks for the reviewing Timothy.
Firstly, the screenshot attached above has nothing to do with the price or quantity, basically we want to ensure that the item that we are performing the "UPDATE" operation exists, so I added the check above. If the item does not exist, we cannot perform the "UPDATE" operation and we will get a 404 not found message.
Similarly, we also want to guarantee that the associated shopcart exists.

Reply:
A Sorry, for this comment, I don't quite get it... Which part were you referring to?

> I understand the test passes but I'm confused about this syntax. Is this testing whether quantity has been set as a local variable? Because when I try to do in a test scripy `if foo: pass ` without foo being set, I get a NameError. So if there is only quantity in the request, surely `if price` would throw an error since we haven't set a variable named price. Sorry I may be missing something very obvious due to my inexperience.

Reply:
lines 339 and 342, ```if quantity: ...``` and ``` if price: ... ```

Reply:
> Thanks for the reviewing Timothy.
> Firstly, the screenshot attached above has nothing to do with the price or quantity, basically we want to ensure that the item that we are performing the "UPDATE" operation exists, so I added the check above. If the item does not exist, we cannot perform the "UPDATE" operation and we will get a 404 not found message.
> Similarly, we also want to guarantee that the associated shopcart exists.

Ah sorry I clicked to add comment on line 339, I didn't mean for the comment to be attached to the section before 339

Reply:
all of my comments here are for lines 339 and 342

Reply:
Great question. That is why I added the comment in the pull request.
In my opinion, we should allow three scenarios:
1. a request body contains only "price" but not "quantity"
2. a request body contains only "quantity" but not "price"
3. a request body contains both "quantity" and "price"

Notice that the variable quantity is extracted from line 322 `quantity = req["quantity"]`
the variable price is extracted from line 327 `price = req["price"]`

Line 339 checks that if we have quantity in our request body.
Line 342 checks that if we have price in our request body. (

I guess the syntax of Python is a little confusing, I should have make it more explicit.
We could replace Line 339 with `if quantity is not None` 
and replace line 342 with `if quantity is not None`,
but they have the same effects as the current code.
Let me know if it is not clear.

Reply:
I agree with everything you wrote, but I'm still confused how e.g. line 342 doesn't throw an error if the request body only has quantity not price. In that case, the if line 323 would be negative, meaning  ```price = req["price"]``` line would not be reached, meaning price is not a defined variable when we get to line 342. Surely that would cause a NameError?

Reply:
> I agree with everything you wrote, but I'm still confused how e.g. line 342 doesn't throw an error if the request body only has quantity not price. In that case, the if line 323 would be negative, meaning `price = req["price"]` line would not be reached, meaning price is not a defined variable when we get to line 342. Surely that would cause a NameError?

In your example, `if foo: pass` throws a NameError I guess you do not define the variable foo,
however, in the code we have right now, price is None if our request body does not contain it.

Reply:
I think you're absolutely right, but this is what I'm confused about. Where was price defined in this code such that it is None instead of being undefined? I'm definitely not understanding something about this sorry.

Reply:
> nk you're absolutely right, but this is what I'm confused about. Where was price defined in this code such that it is None instead of being undefined? I'm definitely not understanding something about this sorry.

Ah, I see where you are confused about. Actually I am not sure about this. I guess it is some Python language design behind the scenes that makes the tests pass but do not throw name errors. Sorry that I cannot give a good answer here.

Reply:
Ah no worries. As long as it works lol I'll ask the Prof. Approved